Fix for skipped turn on load in MP issue

davenye opened the pull request:
After loading an MP game the first turns of human players can often be skipped, meaning the AI gets two turns in a row. I traced the problem down to being a result of the players erroneously being flagged as
having already submitted their ready message for the turn. The fix isn't really a fix as I believe the problem is in the host. Instead I just check for the problem upon the first activation of the local player and
send the unready message if need be. 

In short, HasReceivedTurnComplete was sometimes already returning true after a load so I call sendTurnUnready when it does so their turn doesn't end immediately.

I have played a few games (2 human/6 AI) with the change and have not noticed any negative effects but maybe you can. 

I can probably dig up some save games which exhibit the issue but they will have been using a modded version of the 2/7 release. I didn't figure out why the problem occurs only sometimes.

The issue has been spoken about in:
#2576 


EDIT:
Naturally, I just remembered that I, possibly wrongly, assumed that the problem only occurs when player with dynamic/simultaneous turns. My code will only execute in those modes and changes will still need to be made if sequential turns modes also have the same issue. Should be straight forward though.

LoneGazebo commented:
Interesting! We spent some time way back in the day messing with this and had a hard time tracking the bug down. Good work!

davenye commented:
Do we know if the issue also occurs with sequential MP? If so, the implementation of the fix will probably be a bit different to what I have currently got. Assuming it even fixes the problem in that mode.

davenye commented:
I couldn't get a sequential MP game to exhibit the problem so I reverted to my original, simpler implementation. Re-reading the bug report I saw that it was already assumed to only be a simultaneous/hybrid issue anyway, so all should be well.

LoneGazebo commented:
Oops, didn't mean to merge it. Let me know how your tests go.

raphaell666 commented:
So, how is this fix going? I am having this issue 100% of the time when playing multiplayer, a fix would be simply marvelous. I am very happy to have stumbled upon this here, I had no idea a fix even existed!

LoneGazebo commented:
@davenye never reported back, we'll see what he says. Honestly, though, the easiest solution in the meantime is just to save your game twice before exiting. That removes the issue.

raphaell666 commented:
>@davenye never reported back, we'll see what he says. Honestly, though, the easiest solution in the meantime is just to save your game twice before exiting. That removes the issue.

I will have to try that, thanks a lot for the advice! I will also see if I can figure out how to compile the DLL and will test this fix to help you guys if I am able.

davenye commented:
I have been using my fix as per the submitted pull request during a lot of other development work and testing (prepare for some pull requests!).

I did not investigate any further but I did not notice any issues. I see raphaell666 has noticed some deficiencies though.

LoneGazebo commented:
The two of you should coordinate, make sure you are happy with the changes, as they are big under the hood changes.

davenye commented:
I will look at incorporating their additions to my dev branches.

I did have a fairly in depth look at how saving works. I can tell you all the following:
* The various incantations of ICvEngineUtility1::AutoSave do not appear to do anything but change the filename.
* The Save function in MP games simply copies the "most recent" (determined somehow) autosave and gives it your specified name. As I understand it, Civ 5 originally shipped with only MP autosaves. Their manual save implementation was obviously quickly built on top. Presumably they do not allow the writing of new save files during the turn because it must cause some bad state when loading...
* The "double save" trick only works because of a bug that bypasses the special check they have in SaveMenu.lua that looks if it is MP and performs the copying of the autosave and instead it performs the normal SP code which gets the game to write a actual new file. So given Firaxis when out of their way to prevent mid-turn saves, it stands to reason that it is a bad idea to use the "double save" as a fix. Of course, it may be safe to do under certain circumstances. I have not experienced any such problems except for Active Diplomacy requests are not restored after load (as they are not written out in the first place).

I have some pull requests incoming regarding these.

davenye commented:
Also, if this is considered a big change, can we maybe get a beta for MP like you have done for other big changes in the past?

I have been making changes for the past two months and are much wider reaching than this change. I have been testing but as you know testing (esp. MP testing) takes a long time and I can only do so much. I believe my changes will greatly enhance MP experience and will be well worth the effort incorporating.

I am in the process of making the pull requests for you but it is taking some time. I have them all lined up in branches under "pull/*" on my github if you want to see how much there is to go.

LoneGazebo commented:
Make a new branch for yourself on the main repo. I’ll do a merge once your stuff is up and then do an Mp dll.

raphaell666 commented:
Hey @davenye, your findings regarding saving and autosaving seem to corroborate with mine, as I also mentioned a little bit here: https://github.com/LoneGazebo/Community-Patch-DLL/issues/4166. In addition, I tried documenting my findings there if you could take a look. I have gotten slightly closer to the cause of the issue as I mentioned in the topic, but I could not think of a better fix yet than `gDLL->sendTurnUnready()` unfortunately.

However, I recently had an idea: I am testing treating the game as sequential for only the first turn after load, if it was previously simultaneous. This solves the only issue that I could find with your fix, that is with offline players during the load of the save in pitboss mode still lose their turns. Nonetheless this is considerably more challenging and requires a few more changes in the code, but it avoids the dirty `gDLL->sendTurnUnready()`.

@LoneGazebo I played a couple more games this weekend, using the latest version of the fix in my repo, and everything seems to be working perfectly (except for desyncs, which have nothing to do with this). I wish we had access to the netcode of civ5 so we could try fixing desyncs too. In addition, I am unsure a separate MP DLL would really be required, considering these fixes we are discussing would only activate when loading a MP Simultaneous game (unless @davenye is working on other MP things that I am unaware of).

LoneGazebo commented:
That's fine, but just make sure you're wrapping your code in MP-only specifics. I saw a bunch of pull requests earlier but I'm going to hold them until you're comfortable with the code.

LoneGazebo commented:
re: desyncs, at this point there are no RNG elements left in the DLL - any desyncs going on now aren't related to RNG, so there's either something else afoot that we're not seeing or there's just no hope for desync-free MP.

raphaell666 commented:
>Even though HasReceivedTurnComplete is returning true for the local player on the first turn after load, HasSentTurnComplete is not

This is the same that I have experienced. Both `Player.HasSentTurnComplete()` and `Player.isEndTurn()` return false, while `gDLL->HasReceivedTurnComplete(player.GetID())` returns true. This is why I think the issue is in someway related to the graphical interface. I might very well be wrong though.

>I don't understand why PitBoss has the problem if the issue here is successfully being resolved by fixing the local player only

I believe that all players are marked as having sent (through the interface) that their turn has ended, when loading MP games. If they are not active at the time of the loading, they do not send the `gDLL->SendTurnUnready` of your fix as that only works for active players. Does this makes sense considering what you have experienced?

>that would explain why I think it is the humans that miss the turn, while you think the AI have a double turn

What I know for a fact is that the turn count does not increase when loading MP games and the AI playing twice before the human players. The turn count only increases after the human players play (after all the AI playing twice). This is why I say that what actually happens is that the AI plays twice in one turn, rather than the players losing a turn. The players do not lose a turn as the turn count does not increase until after they play. Makes sense?

>@raphaell666 , this version may appeal more to your sensibilities - davenye@85a5c32

I will test it out once I have time and report back.

raphaell666 commented:
I think I successfully managed to merge those changes by merging the branch into which you commited them, to the master at my repository (still getting used to git and github):
https://github.com/raphaell666/Community-Patch-DLL/commit/087747065f7545b46a6bb1ba2a091d5683e47fc7

I will test these soon.

raphaell666 commented:
@davenye With both fixes applied (SendTurnUnready and this newest one), I tried loading the same autosave I uploaded  in the other reply (this is a pitboss game) and the result was the same as before, inactive players at load still lose their turns (I am Siam, the other player is Portugal):
![load_mp_save](https://user-images.githubusercontent.com/25795774/39154841-1a2b798a-4726-11e8-9d6a-0bb0b20df62a.png)

These are the logs from loading this autosave, if they help:
[Logs.zip](https://github.com/LoneGazebo/Community-Patch-DLL/files/1940141/Logs.zip)

davenye commented:
Cheers, I thought as much but I figured you would still be interested since you were spending time making the original fix more elegant (this is an alternative to the original fix, not an addition).

raphaell666 commented:
Fair enough. I think it's up to @LoneGazebo to decide which fix he thinks is best. I will take a closer look at this new fix (see if I have any new ideas) and test next time with only this newest fix enabled, to be sure it works.

LoneGazebo commented:
Not up to me here, I’m not the tester or the testee. :)

davenye commented:
Yeah, I found that out too when I ran a 2 player MP game. I was a bit hasty with letting you know about that possible fix. It does not work for any more than 1 player.

In other news, I finally ran a PitBoss game today for the first time ever. Unfortunately, I only have enough machines to add a solitary human.

I was able to gather enough information to implement what I believe to fix the double turn issue for PitBoss also. It works for one player (heh) but I am pretty sure it works for any number. Once I have cleaned up my code a little I will send you a link to try it out. I will warn you that it is even less elegant than before! It is an addition to the old fix, not a replacement.

@raphaell666, I incorporated your double turn fix (the one in setTurnActive, before the turn publish) and noted that it is indeed an improvement over my original fix. I am not sure if I messed up my testing but I had not seen the difference in the flashing player icon previously. As a result, I built the PitBoss fix upon your version. I think you should submit your fix as a pull request for the mp_bugfixes branch.

raphaell666 commented:
>Once I have cleaned up my code a little I will send you a link to try it out.

Thank you, I will be waiting. It being elegant or not at this point is irrelevant, let's first get it working. At least with C++ I can help more than with LUA (still have no idea what wizardry you used to understand LUA+XML and make AdvancedSaves, props for that).

>I think you should submit your fix as a pull request for the mp_bugfixes branch.

I think I know how to do that without breaking things up, I will try to do that later today. It is still your fix though, I just improved it a tiny bit. Glad to be of help.

davenye commented:
After going though all this we still have one **major** problem to resolve. I still think the humans are having their turn *skipped*! 🤣 

What is worse is that I am using the same logic you are to come to that conclusion...It is most vexing.

When I load, let's say the turn is 10.
Without the fix, the AI has turn 10, then 11 and then humans get turn 11 (and things progress as expected from there), i.e. humans had turn 10 skipped.
With the fix, the AI has turn 10, then the humans have turn 10 (and things progress as expected from there).

This is all in keeping with how the fix prevents the first human turn from finishing before they have a chance to do anything.

Are you seeing this in both PitBoss and Standard MP? I only ever play Standard and usually the host.

raphaell666 commented:
>Without the fix, the AI has turn 10, then 11 and then humans get turn 11 (and things progress as expected from there), i.e. humans had turn 10 skipped.

What I noticed without the fix is that AI literally plays twice in turn 10, and then the humans get their turn still in turn 10. Only after the humans play it advances to turn 11. I may be mis-remembering. I guess it also depends on which version of these recent changes you are using. I will try deactivating all mods later and verify this again.

>Are you seeing this in both PitBoss and Standard MP? I only ever play Standard and usually the host.

One thing we need to get out of the way is what is actually Pitboss in Civ5. What I usually play is either with an actual dedicated server (which is the actual Pitboss mode imo), or in "Standard MP/Pitboss Internet" or whatever the interface calls it, which is simply standard MP but with the Pitboss option active (this option does not let AI jump into the players place when they are not active/disconnected). 

Choosing "Pitboss" in the game interface menu when hosting the game using the client you will actually play with, is the same as choosing normal "Multiplayer" with the Pitboss option active from my experience.

Recently I noticed that using an dedicated server did not help at all (makes the turn slower even, good job to Firaxis), so I have been playing in standard multiplayer with the Pitboss option active.
